Add functionality to plot bar and barh with string labels (Implement #2516)

umairidris opened the pull request:
This patch is my attempt to add the functionality to pass in strings to label the axis of a bar graph (issue #2516). The patch shouldn't break any existing API. I also center the labels based on the width/height passed for a cleaner more professional look (please view the tests). I am happy to hear any feedback and/or concerns!

Cheers!
Umair

umairidris commented:
One concern I have right now is when the label is very long; it may start to overlap other labels. In my testing this occurs with labels with length over 35. I can add support to space the bars out further based on the length of the max label if this is something that should be supported.

jenshnielsen commented:
The Travis docs build failure is not an issue with this Pull request. Fixed by #4267

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -1960,6 +1965,11 @@ def make_iterable(x):
             if len(height) == 1:
                 height *= nbars
+            if nbars and cbook.is_sequence_of_strings(bottom):
+                ticks_loc = [i + height[i]/2 for i in range(nbars)]
```
I would do this as `[i + h / 2 for i, h in enumerate(height)]`, same for the logic in `bar`

umairidris replied:
Thanks, this looks a lot better now! Fixed via 1627e8e

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -1960,6 +1965,11 @@ def make_iterable(x):
+                ticks_loc = [i + height[i]/2 for i in range(nbars)]
+                self.yaxis.set_ticks(ticks_loc)
+                self.yaxis.set_ticklabels(bottom)
+                bottom = range(nbars)
```
Could you use `np.arange` instead?

umairidris replied:
Done.

tacaswell commented:
:+1: for the idea, I am a bit wary of overloading `bottom` and `left` in this way as now you can either label them _or_ have non-zero baseline, but not both.

umairidris commented:
> :+1: for the idea, I am a bit wary of overloading bottom and left in this way as now you can either label them or have non-zero baseline, but not both.

You're right. To address this I added a check and raised ValueError if bottom (bar) or left (barh) are a list of strings. Perhaps we should more explicitly state this in the docstring as well. If you have any suggestions on how to improve this let me know!

tacaswell commented:
Sorry, I transposed the issue.  Now you can't specify both the labels and non-unit bar spacing which is much worse.

umairidris commented:
Ah, I see what you mean. Perhaps a better way to do this is to not touch left and bottom and instead add an extra kwarg 'labels' or something similar?

The use case I am thinking about is: bar([1,2,3], [1,1,1], labels=['a', 'b', 'c'])

tacaswell commented:
Yes, I think that is a better route.  I would also not move the ticks relative to the bar unless explicitly asked to.  Both `bar*` functions have an `align` kwarg which should still be respected.

tacaswell commented:
@umairidris You should also have a look at how boxplot handles this sort of thing.  iirc `bxp` has a bunch of tick-management logic at the bottom of it.

umairidris commented:
Thanks for the feedback and advice @tacaswell! I have addressed it via commit 8d4e8c3.

A few points I should mention:
- left and bottom are no longer overloaded and can be set to space out the bars however the user wishes. This also nicely handles my earlier concern of overlapping bar labels!

- I wanted to match the 'label' bxpstats key used by bxp. However, label is a kwarg that is already being used by bar. I am not sure what the best name for my added kwarg should be... tick_label? bar_label? text_label?

- I added the ability to pass in the labels as a single string to remain consistent with the rest of the API. You can now do bar(0, 1, tick_label='a') (please see added test).

- I moved the code to set the labels to the bottom of the function. This mimics bxp and also fixes interfering with autoscale (self.autoscale_view() needs to be called before the tick labels are set).

- As requested, bar* align is now respected (see updated tests).

I hope I addressed your concerns. Please let me know if you have any further suggestions!

tacaswell commented:
This looks good, just one code comment.

Can you add an entry in https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new so that this gets advertised?

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -1926,6 +1930,8 @@ def make_iterable(x):
         width = make_iterable(width)
         _bottom = bottom
         bottom = make_iterable(bottom)
+        _tick_label = tick_label
```
I would do this as

```python
label_ticks_flag = not tick_label is None
```

Which I think is a bit clearer.

WeatherGod replied:
tick_label is not None, please
On Mar 29, 2015 11:33 PM, "Thomas A Caswell" <notifications@github.com>
wrote:

> In lib/matplotlib/axes/_axes.py
> <https://github.com/matplotlib/matplotlib/pull/4266#discussion_r27365250>:
>
> > @@ -1926,6 +1930,8 @@ def make_iterable(x):
> >          width = make_iterable(width)
> >          _bottom = bottom
> >          bottom = make_iterable(bottom)
> > +        _tick_label = tick_label
>
> I would do this as
>
> label_ticks_flag = not tick_label is None
>
> Which I think is a bit clearer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4266/files#r27365250>.
>

tacaswell replied:
erm, yes, what @WeatherGod said

umairidris commented:
Done!

@tacaswell This is unrelated but should #4221 also have a what's new entry?

tacaswell commented:
@umairidris This needs to be rebased, not sure why.

umairidris commented:
@tacaswell Done! My branch had the old docstrings (not the new numpy format docstrings) which was causing merge conflicts. All fixed now. :)

Umair I.

tacaswell commented:
Great.  I will merge this as soon as travis passes again.
